### ISSUE TITLE
the compile rule should be smarter about when it can avoid rebuilds.

### ISSUE THREAD

**Issue by zenhack** [capnproto/ekam#28]
In particular, right now I'm doing some print debugging and tracing an issue I'm having in the bridge down into stuff defined in libkj. Adding a simple print statement to a .c++ file should only require re-building the one file, and re-linking anything that uses the resulting object. Instead, ekam is re-compiling a whole bunch of stuff downstream in capnp and in sandstorm.

**Comment by zarvox** [comment]
Is that behavior wrong?  I'm imagining that changes to libkj would dirty the capnp compiler binary, and then any generated code output by the capnp compiler is dirtied (since it was one of the inputs to that rule), and then anything that depends on that generated code is also dirtied, which covers the rest of the stuff in capnp and sandstorm.

**Comment by zenhack** [comment]
Ah, perhaps you're right; I hadn't been thinking about it too carefully.

Quoting Drew Fisher (2020-02-20 03:37:14)

**Comment by kentonv** [comment]
Yes, this is probably what's going on. It is admittedly very annoying.

Maybe `capnpc.ekam-rule` should be adjusted to support specifying an external compiler binary as an environment variable, in which case it won't use the one built by the ekam build itself. You could use that while developing if you know you are going to change files that become part of `capnp` but you do not expect to be modifying the output.

A better solution would be for `ekam` to recognize when the code generated by capnp is identical to what was generated before the change, and then it can skip rebuilding anything that depended on that code. But that's a lot more complicated...
